src/Bicep.Core/Registry/TemplateSpecModuleRegistry.cs

```diff
@@ -0,0 +1,112 @@
+
+        public override ModuleReference? TryParseModuleReference(string reference, out DiagnosticBuilder.ErrorBuilderDelegate? failureBuilder) =>
+            TemplateSpecModuleReference.TryParse(reference, out failureBuilder);
+
+        public override bool IsModuleRestoreRequired(TemplateSpecModuleReference reference) =>
+            !this.fileResolver.FileExists(this.GetModuleEntryPointUri(reference));
+
+        public override Task PublishModule(TemplateSpecModuleReference reference, Stream compiled) => throw new NotSupportedException("Template Spec modules cannot be published.");
+
+        public override Uri? TryGetLocalModuleEntryPointUri(Uri parentModuleUri, TemplateSpecModuleReference reference, out DiagnosticBuilder.ErrorBuilderDelegate? failureBuilder)
+        {
+            failureBuilder = null;
+            return this.GetModuleEntryPointUri(reference);
+        }
+
+        public override async Task<IDictionary<ModuleReference, DiagnosticBuilder.ErrorBuilderDelegate>> RestoreModules(IEnumerable<TemplateSpecModuleReference> references)
+        {
+            var statuses = new Dictionary<ModuleReference, DiagnosticBuilder.ErrorBuilderDelegate>();
+
+            foreach (var reference in references)
+            {
+                using var timer = new ExecutionTimer($"Restore module {reference.FullyQualifiedReference}");
+                try
+                {
+                    var repository = this.repositoryFactory.CreateRepository(reference.EndpointUri, reference.SubscriptionId);
+                    var templateSpec = await repository.FindTemplateSpecByIdAsync(reference.TemplateSpecResourceId);
+
+                    await this.SaveModuleToDisk(reference, templateSpec);
+                }
+                catch (TemplateSpecException templateSpecException)
+                {
+                    statuses.Add(reference, x => x.ModuleRestoreFailedWithMessage(reference.FullyQualifiedReference, templateSpecException.Message));
+                    timer.OnFail(templateSpecException.Message);
+                }
+                catch (Exception exception)
+                {
+                    if (exception.Message is { } message)
+                    {
+                        statuses.Add(reference, x => x.ModuleRestoreFailedWithMessage(reference.FullyQualifiedReference, message));
+                        timer.OnFail($"Unexpected exception {exception}: {message}");
+
+                        return statuses;
+                    }
+
+                    statuses.Add(reference, x => x.ModuleRestoreFailed(reference.FullyQualifiedReference));
+                    timer.OnFail($"Unexpected exception {exception}.");
+                }
+            }
+
+            return statuses;
+        }
+
+        private string GetModuleDirectoryPath(TemplateSpecModuleReference reference) => Path.Combine(
+            this.featureProvider.CacheRootDirectory,
```

Comment:
>CacheRootDirectory [](http://example.com/codeflow?start=33&length=18)

Btw, can you confirm that the test I added for the malformed root path covers templates specs too? (I think it will.)

Reply:
Never mind. I see you had to update the assert, so it works :)